lib/prism/parse_result.rb

```diff
@@ -895,4 +909,46 @@ def initialize(locals, forwarding)
   def self.scope(locals: [], forwarding: [])
     Scope.new(locals, forwarding)
   end
+
+  # Given a Method, UnboundMethod or Proc, use its #source_location to parse the file and return a Prism node representing it.
+  # The returned node will either be a DefNode, LambdaNode or CallNode.
+  # Raises ArgumentError if it cannot be found for any reason.
+  # Only works on Ruby 4+ as it needs #source_location to contain column and end line information.
+  def self.node_for(callable)
+    unless callable.is_a?(Method) || callable.is_a?(UnboundMethod) || callable.is_a?(Proc)
+      raise ArgumentError, 'Prism.node_for requires a Method, UnboundMethod or Proc'
+    end
+    source_location = callable.source_location
+    raise ArgumentError, "#source_location is nil for #{callable}" if source_location.nil?
+    raise ArgumentError, '#source_location does not contain column and end_line, this method only works on Ruby 4+' if source_location.size != 5
+    file, start_line, start_column, end_line, end_column = source_location
+
+    unless File.exist?(file)
+      raise ArgumentError, "#source_location[0] is #{file} but this file does not exist"
+    end
+
+    parse_result = Prism.parse_file(file)
+    unless parse_result.success?
+      raise ArgumentError, "#{file} has syntax errors: #{parse_result.errors_format}"
+    end
+    root = parse_result.value
+    start_offset = parse_result.source.line_and_column_to_byte_offset(start_line, start_column)
+    end_offset = parse_result.source.line_and_column_to_byte_offset(end_line, end_column)
+
+    found = root.breadth_first_search do |node|
```

Comment:
We have a method for this that I think we should use here `Node#tunnel`, which we can then iterate through backward. It's just more efficient than breadth_first_search.

Reply:
Right, I was wondering if we also have a depth-first search which might be faster.
I'll try `tunnel` and see if it's faster. It seems to allocate a `Location` for each visited node though, while this avoids it. Maybe tunnel could use `line_and_column_to_byte_offset` to avoid allocations and have simpler comparisons.

Reply:
```
$ ruby -Ilib -rprism -rbenchmark/ips -e 'm = Prism::YieldNode.instance_method(:===); Benchmark.ips { it.report { Prism.node_for(m) } }'
Without tunnel:
35.855 (± 0.0%) i/s   (27.89 ms/i) -    180.000 in   5.020462s
36.105 (± 0.0%) i/s   (27.70 ms/i) -    183.000 in   5.068871s
With tunnel:
37.073 (± 0.0%) i/s   (26.97 ms/i) -    186.000 in   5.017522s
37.179 (± 0.0%) i/s   (26.90 ms/i) -    186.000 in   5.002992s
```
It a tiny bit faster, and using tunnel is very easy, so done.